test/unit/math/rev/mat/prob/categorical_logit_glm_lpmf_test.cpp

```diff
@@ -0,0 +1,569 @@
+    EXPECT_NEAR(alpha1[i].adj(), alpha2[i].adj(), eps);
+  }
+}
+
+TEST(ProbDistributionsCategoricalLogitGLM, y_broadcasting) {
+  double eps = 1e-13;
+  const size_t N_instances = 5;
+  const size_t N_attributes = 2;
+  const size_t N_classes = 3;
+  Matrix<var, Dynamic, Dynamic> x1(N_instances, N_attributes),
+      x2(N_instances, N_attributes);
+  x1 << -12, 46, -42, 24, 25, 27, -14, -11, 5, 18;
+  x2 << -12, 46, -42, 24, 25, 27, -14, -11, 5, 18;
+  Matrix<var, Dynamic, Dynamic> beta1(N_attributes, N_classes),
+      beta2(N_attributes, N_classes);
+  beta1 << 0.3, 2, 0.4, -0.1, -1.3, 1;
+  beta2 << 0.3, 2, 0.4, -0.1, -1.3, 1;
+  Matrix<var, Dynamic, 1> alpha1(N_classes), alpha2(N_classes);
+  alpha1 << 0.5, -2, 4;
+  alpha2 << 0.5, -2, 4;
+
+  for (int y_scal = 1; y_scal <= N_classes; y_scal++) {
+    vector<int> y(N_instances, y_scal);
+    var res1 = categorical_logit_glm_lpmf(y_scal, x1, alpha1, beta1);
+    var res2 = categorical_logit_glm_lpmf(y, x2, alpha2, beta2);
+    (res1 + res2).grad();
+    EXPECT_NEAR(res1.val(), res2.val(), eps);
+    for (int i = 0; i < N_attributes; i++) {
+      for (int j = 0; j < N_instances; j++) {
+        EXPECT_NEAR(x1(j, i).adj(), x2(j, i).adj(), eps);
+      }
+      for (int j = 0; j < N_classes; j++) {
+        EXPECT_NEAR(beta1(i, j).adj(), beta2(i, j).adj(), eps);
+      }
+    }
+    for (int i = 0; i < N_classes; i++) {
+      EXPECT_NEAR(alpha1[i].adj(), alpha2[i].adj(), eps);
+    }
+
+    stan::math::set_zero_all_adjoints();
+
+    res1 = categorical_logit_glm_lpmf<true>(y_scal, x1, alpha1, beta1);
+    res2 = categorical_logit_glm_lpmf<true>(y, x2, alpha2, beta2);
+    (res1 + res2).grad();
+    EXPECT_NEAR(res1.val(), res2.val(), eps);
+    for (int i = 0; i < N_attributes; i++) {
+      for (int j = 0; j < N_instances; j++) {
+        EXPECT_NEAR(x1(j, i).adj(), x2(j, i).adj(), eps);
+      }
+      for (int j = 0; j < N_classes; j++) {
+        EXPECT_NEAR(beta1(i, j).adj(), beta2(i, j).adj(), eps);
+      }
+    }
+    for (int i = 0; i < N_classes; i++) {
+      EXPECT_NEAR(alpha1[i].adj(), alpha2[i].adj(), eps);
+    }
+  }
+}
+
+TEST(ProbDistributionsCategoricalLogitGLM,
+     glm_matches_categorical_logit_vars_big) {
+  double eps = 1e-11;
+  const size_t N_instances = 89;
+  const size_t N_attributes = 23;
+  const size_t N_classes = 11;
+  vector<int> y(N_instances);
+  for (int i = 0; i < N_instances; i++) {
+    y[i] = Matrix<int, Dynamic, 1>::Random(1)[0] % N_classes + 1;
+  }
+  MatrixXd x_double = MatrixXd::Random(N_instances, N_attributes);
+  Matrix<var, Dynamic, Dynamic> x1 = x_double, x2 = x_double;
+  MatrixXd beta_double = MatrixXd::Random(N_attributes, N_classes);
+  Matrix<var, Dynamic, Dynamic> beta1 = beta_double, beta2 = beta_double;
+  RowVectorXd alpha_double = RowVectorXd::Random(N_classes);
+  Matrix<var, Dynamic, 1> alpha1 = alpha_double, alpha2 = alpha_double;
+
+  var res1 = categorical_logit_glm_simple_lpmf<false>(y, x1, alpha1, beta1);
```

Comment:
What's simple? the literal `softmax(inv_logit(foo))`, and you're testing against your new implementation?

Reply:
The function `categorical_logit_glm_simple_lpmf()` is implemented at start of the file. It is a simple implemtation of the new GLM - it uses autodiff instead of analytically derived gradients.